kv.go

```diff
@@ -10,83 +10,159 @@ import "context"
+// it can be provided when creating an [Iterator] from a [Store].
 type IterOptions struct {
-	// Prefix iteration, all keys *must* contain
-	// the designated prefix. Prefix takes precendent
-	// over range (start/end). If Prefix is nil, range
-	// will be used (regardless if start/end is nil)
+	// Prefix iteration, only keys beginning with the designated prefix
+	// with the given prefix will be yielded.
+	//
+	// Keys exactly matching the provided `Prefix` value will not be
+	// yielded.
+	//
+	// Providing a Prefix value should cause the Start and End options
+	// to be ignored, although this is currently untested:
+	// https://github.com/sourcenetwork/corekv/issues/35
 	Prefix []byte
 
-	// Range iteration from start (inclusive) to end explusive.
-	// Start must be lexigraphically less
-	// then end, unless `nil` is used.
+	// If Prefix is nil, and Start is provided, the iterator will
+	// only yield items with a key lexographically greater than or
+	// equal to this value.
 	//
-	// Start defined as nil will iterate from the first
-	// key in the store.
-	// End defined as nil will iterate till the last key
-	// in the store.
+	// Providing an `End` value equal to or smaller than this value
+	// will result in undefined behaviour:
+	// https://github.com/sourcenetwork/corekv/issues/32
 	Start []byte
-	End   []byte
 
-	// Reverse the direction of the iteration.
+	// If Prefix is nil, and Start is provided, the iterator will
+	// only yield items with a key lexographically smaller than this
+	// value.
+	//
+	// Providing an End value equal to or smaller than Start
+	// will result in undefined behaviour:
+	// https://github.com/sourcenetwork/corekv/issues/32
+	End []byte
+
+	// Reverse the direction of the iteration, returning items in
+	// lexographically descending order of their keys.
 	Reverse bool
 
-	// Only iterate through keys. Calling Value() on the
-	// iterator *must* return nil (no error).
+	// Only iterate through keys. Calling Value on the
+	// iterator will return nil and no error.
+	//
+	// This option is currently untested:
+	// https://github.com/sourcenetwork/corekv/issues/34
+	//
+	// It is very likely ignored for the memory store iteration:
+	// https://github.com/sourcenetwork/corekv/issues/33
 	KeysOnly bool
 }
 
+// Reader contains read-only functions for interacting with a store.
 type Reader interface {
+	// Get returns the value at the given key.
+	//
+	// If no item with the given key is found, nil, and an [ErrNotFound]
+	// error will be returned.
 	Get(ctx context.Context, key []byte) ([]byte, error)
+
+	// Has returns true if an item at the given key is found, otherwise
+	// will return false.
 	Has(ctx context.Context, key []byte) (bool, error)
+
+	// Iterator returns a read-only iterator using the given options.
 	Iterator(ctx context.Context, opts IterOptions) Iterator
 }
 
+// Writer contains functions for mutating stuff stored within a store.
```

Comment:
suggestion: Mutating *stuff* -> Mutating *values*

Reply:
I do disagree, as it is not just for mutating `values`, but anything in the store (e.g. adding/removing keys). 'Stuff' was a deliberate choice on my part.

Reply:
There are only two things stored in our stores and you can't have one without the other. They are key and value. So if you delete a value you are effectively deleting the kv pair. You are not just adding a key, you are setting a kv pair. It's not *stuff*. We can be more precise here. I don't see the reason for the ambiguity.

Reply:
Ah I see what you mean, and the lack of a `Move`-like func on the interface supports this thought atm, and transactions/iterators (tracked/mutated in store state) are not impacted either by this particular interface.  I'll use the word `value`.

 - [ ] Mutating stuff -> Mutating values